### DISCUSSION TITLE
How to ensure the integrity of project references when the namespace between layers is the same?

### DISCUSSION THREAD

**Discussion by HHobeck** [realvizu/NsDepCop#73] [General]
Hi there.

In my projects I'm following the domain driven design paradigma and name the system namespaces without including any technical aspects like Service, BusinessLogic and Repository (a classic 3-layer architecture). Of course the technical implementations (layers) of the domain models are isolated via different assemblies but living within the same namespace.

For example:
- MyCompany.Service.dll references MyCompany.ServiceContract.dll and MyCompanyRepositoryContract.dll
- MyCompany.Repository.dll references MyCompanyRepositoryContract.dll

**To my question: How would you ensure the integrity of the layers if you want prohibit a reference from, for example, the layer MyCompany.Repository.dll to MyCompany.ServiceContract.dll?**

**Comment by realvizu** [comment]
Let me rephrase your question to see if I got it. So you use namespaces and assemblies as two orthogonal dimensions to divide up your application: you use namespaces for the business areas and assemblies for the technical layers, right?

Unfortunately this structure doesn't make it possible to use NsDepCop for controlling dependencies between technical layers, because those are not identified by namespaces, but by assemblies. And NsDepCop checks dependencies only between namespaces.

I see a couple of options though.

(1) Use some other tool for controlling dependencies between assemblies. 
E.g. https://github.com/TNG/ArchUnitNET lets you write tests that check architectural rules. So you could use it to test that there are no prohibited assembly references.
This seems like the simplest solution.

(2) Reorganize namespaces so you can use them to represents both domain (business area) and technical (layer) aspects.
E.g. you could have namespaces like: 
* MyApp.ProductCatalog.ServiceContract, MyApp.ProductCatalog.Service, MyApp.ProductCatalog.RepositoryContract, etc.
* MyApp.OrderProcessing.ServiceContract, MyApp.OrderProcessing.Service, MyApp.OrderProcessing.RepositoryContract, etc.
* etc.

This approach has the benefit that you can define much more fine grained dependency rules between namespaces than it would be possible with just assemblies.
E.g.
* MyApp.OrderProcessing.Service can depend on MyApp.OrderProcessing.ServiceContract, because it implements it.
* MyApp.OrderProcessing.Service can depend on MyApp.ProductCatalog.ServiceContract, because it can use it.
* But you don't want to allow all things in the Service layer to use all things in the ServiceContract layer.

I've got a blog post that further elaborates on this topic: https://realvizu.com/2018/02/03/implementing-ddd-bounded-contexts-with-csharp-namespaces-and-dependency-rules/

> As a side note: you could have many sub-namespaces under each layer-identifying namespace. 
> E.g.
> * MyApp.ProductCatalog.Service.Search
> * MyApp.ProductCatalog.Service.Pricing
> * etc.

**Reply by HHobeck** [reply]
> Let me rephrase your question to see if I got it. So you use namespaces and assemblies as two orthogonal dimensions to divide up your application: you use namespaces for the business areas and assemblies for the technical layers, right?

Yes that is correct.

> (1) Use some other tool for controlling dependencies between assemblies.
> E.g. https://github.com/TNG/ArchUnitNET lets you write tests that check architectural rules. So you could use it to test that there are no prohibited assembly references.
> This seems like the simplest solution.
> 
> (2) Reorganize namespaces so you can use them to represents both domain (business area) and technical (layer) aspects.
> E.g. you could have namespaces like:
> 
> MyApp.ProductCatalog.ServiceContract, MyApp.ProductCatalog.Service, MyApp.ProductCatalog.RepositoryContract, etc.
> MyApp.OrderProcessing.ServiceContract, MyApp.OrderProcessing.Service, MyApp.OrderProcessing.RepositoryContract, etc.
> etc.

Thank you very much providing some solutions for my use case. I need to take a look into the option (1) and verify how complex it might be to create such tests. Anyway I would prefer a solution with static code analysis which is the reason why I like NsDepCop. The option (2) might be also an idea. The reason why I don't like to specify technical layers in the namespace is because of the Domain Driven Design paradigm. It should be very easy to change the project structure and move the child domain to another assembly. From the business perspective such a refactoring initiative makes no different and is equivalent.

Let's say if you have a repository with name ProductRepository in the namespace MyApp.ProductCatalog then you can place it:
- in the project MyApp.ProudctCatalog.Repositories on the root level or
- in MyApp.Repositories in the subfolder ProductCatalog

Of course both things needs to be ensured the layer integrity and the domain integrity itself.

**Reply by HHobeck** [reply]
Do you think checking the DLL reference integrity might fit in this project? E.g. it could be implemented (with your guidance) as an additional feature from my person.

**Reply by realvizu** [reply]
Enforcing assembly dependency rules would be a nice feature.
Do you have a proposal on how would you extend the config file schema for that? (While being backward compatible, of course.)

**Reply by HHobeck** [reply]
I would say:
```xml
<?xml version="1.0" encoding="utf-8"?>
<NsDepCopConfig>
    <Allowed From="*" To="*" />
    <Disallowed From="A" To="B" />
    <AllowedAssemblyDependency From="*" To="*" />
    <DisallowedAssemblyDependency From="*.RepositoryContract" To="*.ServiceContract" />
    <DisallowedAssemblyDependency From="*.RepositoryContract" To="*.Service" />
    <DisallowedAssemblyDependency From="*.Repository" To="*.ServiceContract" />
    <DisallowedAssemblyDependency From="*.Repository" To="*.Service" />
</NsDepCopConfig>
```

**Reply by realvizu** [reply]
If we adopt the "disallowed-by-default" approach for assembly dependencies, similar to how we handle namespace dependencies (where everything is disallowed unless explicitly permitted), we should ensure backward compatibility. This can be achieved by introducing a configuration attribute to disable assembly dependency checking by default. This ensures that existing configurations without any assembly dependency rules do not start reporting unexpected issues.

For instance:

```xml
<NsDepCopConfig CheckAssemblyDependencies="true"> <!-- the default value of CheckAssemblyDependencies should be false -->
``` 

For the new configuration elements, I suggest shorter names like `AllowedAssembly` and `DisallowedAssembly` to reduce verbosity.

Some thoughts about the implementation:
* The entry point is the NsDepCopAnalyzer class, instantiated and called by the Roslyn AnalyzerExecutor.
  * You can use the NsDepCop.Vsix project to experiment with it and debug it.
  * Unfortunately there are no unit tests for NsDepCopAnalyzer yet (see issue #80).
  * In NsDepCopAnalyzer.StartAnalysisForCompilation you receive the CompilationStartAnalysisContext which should have all info that you need for checking the current project's assembly dependencies:
    * compilationStartContext.Compilation.AssemblyName
	* compilationStartContext.Compilation.ReferencedAssemblyNames
* You'll have to create a new DiagnosticDescriptor in the DiagnosticDefinitions class for illegal assembly references. Please also add the new diagnostic to the doc/Diagnostics.md file.
* Currently the dependency validation logic is implemented by TypeDependencyEnumerator (finding dependencies) and CachingTypeDependencyValidator (deciding whether a dependency is valid). You will need something similar for assemblies, but it will have to work for a compilation (project), not for a syntax node. 
* You may or may not be able to reuse some part of the rule matching logic implemented in TypeDependencyValidator, because the current implementation is quite specific for types/namespaces and I don't know wether it can by generalized for assembly dependencies too.
* Note that the AssemblyDependencyValidator won't need caching (unlike the CachingTypeDependencyValidator), because the dependency validation happens at the C# project level, so there can be no multiple references to the same assembly.
* You'll also have to extend the IDependencyRules interface with the assembly dependency rules (and its implementation in AnalyzerConfig and the config parsing logic in XmlConfigParser and AnalyzerConfigBuilder).

Hope this helps getting started. Feel free to reach out with questions or for further guidance. I’m happy to assist!

**Reply by HHobeck** [reply]
> If we adopt the "disallowed-by-default" approach for assembly dependencies, similar to how we handle namespace dependencies (where everything is disallowed unless explicitly permitted), we should ensure backward compatibility. This can be achieved by introducing a configuration attribute to disable assembly dependency checking by default. This ensures that existing configurations without any assembly dependency rules do not start reporting unexpected issues.

I agree with you. Good point.

> For the new configuration elements, I suggest shorter names like AllowedAssembly and DisallowedAssembly to reduce verbosity.

Okay no problem.

> The entry point is the NsDepCopAnalyzer class, instantiated and called by the Roslyn AnalyzerExecutor.
> * You can use the NsDepCop.Vsix project to experiment with it and debug it.

Actually I have some problems with the NsDepCop.Vsix project with Visual Studio 2022 and Windows 11 (see [here](https://learn.microsoft.com/en-us/answers/questions/846036/vsix-installer-does-not-work-on-windows-11)). Even if I update the nuget packages I cannot use it, because when installing it VS tells me that: VSIXInstaller.NoApplicableSKUsException: This extension is not installable on any currently installed products.

Anyway, I have found a workaround when adding the project directly to my test solution:

```xml
<ItemGroup>
  <ProjectReference Include="..\NsDepCop\source\NsDepCop.Analyzer\NsDepCop.Analyzer.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
</ItemGroup>
```

> * Unfortunately there are no unit tests for NsDepCopAnalyzer yet (see issue [Write unit tests for NsDepCopAnalyzer #80](https://github.com/realvizu/NsDepCop/issues/80)).

Good to know. Thank you for the hint.

> * In NsDepCopAnalyzer.StartAnalysisForCompilation you receive the CompilationStartAnalysisContext which should have all info that you need for checking the current project's assembly dependencies:
>   
>   * compilationStartContext.Compilation.AssemblyName
>   * compilationStartContext.Compilation.ReferencedAssemblyNames

It might be an idea to use the `RegisterCompilationAction` instead because I can use the `ReportDiagnostic` on project level. It's more intuitive in my opinion and has the same outcome:

```yaml
public override void Initialize(AnalysisContext analysisContext)
{
    analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
    analysisContext.EnableConcurrentExecution();

    analysisContext.RegisterCompilationAction(context => context.ReportDiagnostic(..));
}
```

**Reply by HHobeck** [reply]
I have created a pull-request. Please take a look: https://github.com/realvizu/NsDepCop/pull/81

Following points are open:
- ~~renaming from AllowedAssemblyDependency to AllowedAssembly~~
- ~~renaming from DisallowedAssemblyDependency to DisallowedAssembly~~
- ~~creating CheckAssemblyDependencies attribute in NsDepCopConfig~~
- ~~addding the new diagnostic to the doc/Diagnostics.md file~~
- ~~writing unit tests~~

**Reply by HHobeck** [reply]
Done. Wish you a happy new year! ;)

**Reply by realvizu** [reply]
Thanks a lot for your work, I'll review it soon.
Happy new year to you too!

**Reply by realvizu** [reply]
PR merged, I will release the new nuget version soon.
Thanks again!

**Reply by HHobeck** [reply]
@realvizu: Please see the discussion https://github.com/dotnet/roslyn/issues/50574#issuecomment-2600752396

**Comment by realvizu** [comment]
Nuget released: https://www.nuget.org/packages/NsDepCop/2.5.0
